### ISSUE TITLE
Documentation

### ISSUE THREAD

**Issue by FinnStokes** [ILAFF/ILAFF#7]
We should generate documentation from the docstrings in our code. The most popular option seems to be [Sphinx](https://www.sphinx-doc.org/en/master/), but there is also [pdoc](https://pdoc3.github.io/pdoc/), which is apparently easier to set up for simpler projects. I think given the build infrasructure we already have set up, Sphinx shouldn't be too much extra burden.

I lean towards using [Google](http://google.github.io/styleguide/pyguide.html#Comments) or [NumPy](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt)-style docstrings, which would mean using the inbuilt [Napoleon](https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html) extension if we use Sphinx.

Once we get it set up, it would be good to add the documentation generation to the CI and [upload as build artifacts](https://help.github.com/en/actions/language-and-framework-guides/using-python-with-github-actions#packaging-workflow-data-as-artifacts) and/or push to [GitHub Pages](https://stackoverflow.com/questions/57989790/using-github-actions-to-publish-documentation) or [Read the Docs](https://readthedocs.org/)

**Comment by achambers8472** [comment]
I personally prefer the Google style - the numpy header underlines add a lot of vertical space and it's a pain having to get their length correct - a pain with restructured text generally

**Comment by achambers8472** [comment]
I can have a crack at setting this up :) I've done the Sphinx/Napoleon thing before, but I've never tried to push the generated doc's as an artifact.

**Comment by achambers8472** [comment]
I've pushed an initial implementation to feature/documentation (I don't know how to reference a branch here).  To build the docs, you can just go into `source/` and run `make html`.

One thing that's interesting to discuss is the use of `sphinx-apidoc` (not to be confused with `sphinx-autodoc`.  `sphinx-autodoc` is great and I love it, but I don't like `sphinx-apidoc`'s layout, personally I like to create the source-tree manually.  But happy to be overruled ;)

**Comment by achambers8472** [comment]
Will have a crack at uploading documentation as an artifact this afternoon :)

**Comment by achambers8472** [comment]
The initial version of this has been merged into master, and a build pipeline pushing to readthedocs has been setup: https://ilaff.readthedocs.io/en/latest/

I'm still learning about the setup, but basically what happens is that on readthedocs, there is a trigger for building fresh documentation, in this case updates to master.  I can add anyone with a readthedocs account as admins to the ilaff documentation, just ping me!

**Comment by FinnStokes** [comment]
So `sphinx-autodoc` generates documentation from the sources you tell it to in your manually configured documentation tree, whereas `sphinx-apidoc` automatically builds a documentation tree using `sphinx-autodoc` by parsing your package?

**Comment by achambers8472** [comment]
Exactly!  It's pretty easy to play around with, you just do something like `sphinx-apidoc ../src/ilaff`.  Maybe I just don't know how to configure it properly, but with the default settings it makes the package tree very complicated to traverse?  As in, you go ilaff -> ilaff submodules -> ilaff.units -> ilaff.units contents -> to get to anything about any of your ilaff.units classes, for example.

**Comment by rsteelephys** [comment]
Looks like I'll be taking on some of the documentation. A lot of the documentation and examples for units is currently as a docstring in the top of  the module, so I'll be moving it/ trying to rearrange the documentation to  more of a readthedocs tutorial format.

So @achambers8472 I guess we should talk about adding me to readthedocs at some point

**Comment by achambers8472** [comment]
Hi @rsteelephys , absolutely, if you give me the email address associated with your readthedocs account, I can make you an admin for the project!
